File: sdk/python/src/veoveo_mcp/internal_auth.py
"""Gateway internal token verification for hosted Python MCP servers.

Mirrors the Rust `mcp-contract` internal-auth module: the gateway alone signs
short-lived Ed25519 (EdDSA) identity assertions; hosted servers receive a
public JWKS trust bundle, require a `kid`, and never hold the private key.
"""

from __future__ import annotations

import base64
import json
from datetime import datetime, timezone
from typing import Any, Awaitable, Callable

import jwt as pyjwt
from pydantic import ValidationError

from .contract.identity import GatewayInternalIdentity, InvocationAuthority, Principal

Scope = dict[str, Any]
Receive = Callable[[], Awaitable[dict[str, Any]]]
Send = Callable[[dict[str, Any]], Awaitable[None]]
AsgiApp = Callable[[Scope, Receive, Send], Awaitable[None]]

IDENTITY_SCOPE_KEY = "veoveo.internal_identity"
BEARER_SCOPE_KEY = "veoveo.forwarded_bearer"

_REQUIRED_CLAIMS = ["exp", "iss", "aud", "sub", "iat", "nbf", "jti"]


class InternalTokenError(Exception):
    pass


class GatewayInternalTrustBundle:
    """Validated Ed25519 verification keys, indexed by `kid`."""

    def __init__(self, jwks: dict[str, Any]) -> None:
        keys = jwks.get("keys")
        if not isinstance(keys, list) or not keys:
            raise InternalTokenError("internal trust JWKS contains no keys")
        self._keys: dict[str, pyjwt.PyJWK] = {}
        for jwk in keys:
            _validate_verification_jwk(jwk)
            key_id = _validate_key_id(jwk.get("kid"))
            if key_id in self._keys:
                raise InternalTokenError(
                    f"internal trust JWKS contains duplicate kid `{key_id}`"
                )
            self._keys[key_id] = pyjwt.PyJWK(jwk, algorithm="EdDSA")

    @classmethod
    def from_json(cls, value: str) -> "GatewayInternalTrustBundle":
        try:
            jwks = json.loads(value)
        except json.JSONDecodeError as error:
            raise InternalTokenError(f"invalid internal trust JWKS: {error}") from error
        return cls(jwks)

    def key(self, key_id: str) -> pyjwt.PyJWK:
        key = self._keys.get(key_id)
        if key is None:
            raise InternalTokenError(
                f"internal token references unknown kid `{key_id}`"
            )
        return key

    def key_ids(self) -> list[str]:
        return list(self._keys)


def _validate_key_id(value: Any) -> str:
    if (
        not isinstance(value, str)
        or not value
        or value.strip() != value
        or any(ch for ch in value if ch < " " or ch == "\x7f")
    ):
        raise InternalTokenError("internal signing key id is empty or invalid")
    return value


def _validate_verification_jwk(jwk: dict[str, Any]) -> None:
    if jwk.get("alg") != "EdDSA":
        raise InternalTokenError(
            "internal trust JWKS must contain Ed25519 verification keys"
        )
    use = jwk.get("use")
    if use is not None and use != "sig":
        raise InternalTokenError(
            "internal trust JWKS must contain Ed25519 verification keys"
        )
    key_ops = jwk.get("key_ops")
    if key_ops is not None and (not key_ops or any(op != "verify" for op in key_ops)):
        raise InternalTokenError(
            "internal trust JWKS must contain Ed25519 verification keys"
        )
    if jwk.get("kty") != "OKP" or jwk.get("crv") != "Ed25519":
        raise InternalTokenError(
            "internal trust JWKS must contain Ed25519 verification keys"
        )
    x = jwk.get("x")
    try:
        decoded = base64.urlsafe_b64decode(x + "=" * (-len(x) % 4)) if x else b""
    except (TypeError, ValueError):
        decoded = b""
    if len(decoded) != 32:
        raise InternalTokenError(
            "internal trust JWKS must contain Ed25519 verification keys"
        )


class GatewayInternalTokenVerifier:
    def __init__(
        self,
        issuer: str,
        audiences: list[str],
        trust_bundle: GatewayInternalTrustBundle,
    ) -> None:
        if not audiences:
            raise InternalTokenError("internal token verifier requires an audience")
        self.issuer = issuer
        self.audiences = audiences
        self.trust_bundle = trust_bundle

    @classmethod
    def for_server(
        cls, issuer: str, server_slug: str, trust_bundle: GatewayInternalTrustBundle
    ) -> "GatewayInternalTokenVerifier":
        return cls(issuer, [server_slug], trust_bundle)

    def verify(self, bearer_token: str) -> GatewayInternalIdentity:
        try:
            header = pyjwt.get_unverified_header(bearer_token)
        except pyjwt.InvalidTokenError as error:
            raise InternalTokenError(
                f"internal token JWT validation failed: {error}"
            ) from error
        if header.get("alg") != "EdDSA":
            raise InternalTokenError(
                f"internal token algorithm `{header.get('alg')}` is not EdDSA"
            )
        key_id = header.get("kid")
        if key_id is None:
            raise InternalTokenError("internal token or trust key is missing kid")
        key = self.trust_bundle.key(key_id)
        try:
            claims = pyjwt.decode(
                bearer_token,
                key,
                algorithms=["EdDSA"],
                issuer=self.issuer,
                audience=self.audiences,
                leeway=0,
                options={"require": _REQUIRED_CLAIMS},
            )
        except pyjwt.InvalidTokenError as error:
            raise InternalTokenError(
                f"internal token JWT validation failed: {error}"
            ) from error
        server = claims.get("server")
        if server not in self.audiences:
            raise InternalTokenError(
                f"internal token audience mismatch: got `{server}`"
            )
        try:
            actor = Principal.model_validate(claims["actor"])
            authority = InvocationAuthority.model_validate(claims["authority"])
        except (KeyError, ValidationError) as error:
            raise InternalTokenError(
                f"internal token identity is invalid: {error}"
            ) from error
        if claims["sub"] != actor.id:
            raise InternalTokenError(
                "internal token subject does not match embedded actor"
            )
        return GatewayInternalIdentity(
            issuer=claims["iss"],
            profile=claims["profile"],
            server=server,
            actor=actor,
            authority=authority,
            jwt_id=claims["jti"],
            issued_at=_timestamp(claims["iat"], "iat"),
            not_before=_timestamp(claims["nbf"], "nbf"),
            expires_at=_timestamp(claims["exp"], "exp"),
        )


def _timestamp(value: Any, claim: str) -> datetime:
    try:
        return datetime.fromtimestamp(int(value), tz=timezone.utc)
    except (TypeError, ValueError, OSError) as error:
        raise InternalTokenError(
            f"internal token claim `{claim}` has invalid timestamp `{value}`"
        ) from error


def bearer_from_header(header: str) -> str:
    scheme, _, token = header.partition(" ")
    if not scheme.lower() == "bearer":
        raise InternalTokenError("authorization scheme must be Bearer")
    if not token or any(ch.isspace() for ch in token):
        raise InternalTokenError("bearer token contains invalid whitespace")
    return token


class InternalAuthMiddleware:
    """Reject requests without a valid gateway assertion; stash the identity.

    Downstream layers read `IDENTITY_SCOPE_KEY` (the parsed
    `GatewayInternalIdentity`) and `BEARER_SCOPE_KEY` (the forwarded bearer)
    from the ASGI scope.
    """

    def __init__(
        self,
        app: AsgiApp,
        verifier: GatewayInternalTokenVerifier,
        logger: Callable[[str], None] | None = None,
    ) -> None:
        self.app = app
        self.verifier = verifier
        self.logger = logger

    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
        if scope["type"] != "http":
            await self.app(scope, receive, send)
            return
        header = None
        for name, value in scope.get("headers", []):
            if name.decode("latin-1").lower() == "authorization":
                header = value.decode("latin-1")
                break
        try:
            if header is None:
                raise InternalTokenError("missing internal authorization")
            token = bearer_from_header(header)
            identity = self.verifier.verify(token)
        except InternalTokenError as error:
            if self.logger is not None:
                self.logger(f"rejected MCP request: {error}")
            body = b"invalid gateway authorization"
            await send(
                {
                    "type": "http.response.start",
                    "status": 401,
                    "headers": [
                        (b"content-type", b"text/plain; charset=utf-8"),
                        (b"content-length", str(len(body)).encode()),
                    ],
                }
            )
            await send({"type": "http.response.body", "body": body})
            return
        scope[IDENTITY_SCOPE_KEY] = identity
        scope[BEARER_SCOPE_KEY] = token
        await self.app(scope, receive, send)

File: servers/chart-mcp/internal-auth.mjs
import {
  createPublicKey,
  verify as verifySignature,
} from "node:crypto";

const INTERNAL_TOKEN_ISSUER = "veoveo-internal";

export function loadInternalTokenVerifier(rawJwks, audience) {
  if (typeof audience !== "string" || audience.length === 0) {
    throw new Error("internal token verifier requires an audience");
  }
  if (typeof rawJwks !== "string" || rawJwks.trim().length === 0) {
    throw new Error("VEOVEO_INTERNAL_TRUST_JWKS is required");
  }
  let bundle;
  try {
    bundle = JSON.parse(rawJwks);
  } catch (error) {
    throw new Error(`VEOVEO_INTERNAL_TRUST_JWKS is invalid JSON: ${String(error)}`);
  }
  if (!Array.isArray(bundle.keys) || bundle.keys.length === 0) {
    throw new Error("VEOVEO_INTERNAL_TRUST_JWKS must contain verification keys");
  }
  const keys = new Map();
  for (const jwk of bundle.keys) {
    if (
      jwk == null ||
      typeof jwk !== "object" ||
      typeof jwk.kid !== "string" ||
      jwk.kid.length === 0 ||
      jwk.kty !== "OKP" ||
      jwk.crv !== "Ed25519" ||
      (jwk.alg != null && jwk.alg !== "EdDSA") ||
      (jwk.use != null && jwk.use !== "sig") ||
      (jwk.key_ops != null &&
        (!Array.isArray(jwk.key_ops) ||
          jwk.key_ops.length === 0 ||
          jwk.key_ops.some((operation) => operation !== "verify"))) ||
      typeof jwk.x !== "string" ||
      Buffer.from(jwk.x, "base64url").length !== 32
    ) {
      throw new Error("internal trust JWKS contains an unsupported verification key");
    }
    if (keys.has(jwk.kid)) {
      throw new Error(`internal trust JWKS repeats kid ${jwk.kid}`);
    }
    keys.set(
      jwk.kid,
      createPublicKey({
        key: jwk,
        format: "jwk",
      }),
    );
  }
  return (token) => verifyInternalToken(token, keys, audience);
}

function verifyInternalToken(token, keys, audience) {
  const segments = token.split(".");
  if (segments.length !== 3 || segments.some((segment) => segment.length === 0)) {
    throw new Error("internal token is not a compact JWT");
  }
  let header;
  let claims;
  try {
    header = JSON.parse(Buffer.from(segments[0], "base64url").toString("utf8"));
    claims = JSON.parse(Buffer.from(segments[1], "base64url").toString("utf8"));
  } catch (error) {
    throw new Error(`internal token has invalid JSON: ${String(error)}`);
  }
  if (header.alg !== "EdDSA" || typeof header.kid !== "string") {
    throw new Error("internal token requires EdDSA and kid");
  }
  const key = keys.get(header.kid);
  if (
    key == null ||
    !verifySignature(
      null,
      Buffer.from(`${segments[0]}.${segments[1]}`),
      key,
      Buffer.from(segments[2], "base64url"),
    )
  ) {
    throw new Error("internal token signature is invalid");
  }
  const now = Math.floor(Date.now() / 1000);
  if (
    claims.iss !== INTERNAL_TOKEN_ISSUER ||
    claims.aud !== audience ||
    claims.server !== audience ||
    typeof claims.sub !== "string" ||
    claims.sub.length === 0 ||
    typeof claims.jti !== "string" ||
    claims.jti.length === 0 ||
    typeof claims.profile !== "string" ||
    claims.profile.length === 0 ||
    claims.actor == null ||
    typeof claims.actor !== "object" ||
    claims.actor.id !== claims.sub ||
    claims.authority == null ||
    typeof claims.authority !== "object" ||
    !Number.isInteger(claims.iat) ||
    !Number.isInteger(claims.nbf) ||
    !Number.isInteger(claims.exp) ||
    claims.iat > now ||
    claims.nbf > now ||
    claims.exp <= now
  ) {
    throw new Error("internal token claims are invalid");
  }
  return claims;
}

export function requireInternalIdentity(request, response, verifyToken) {
  const authorization = request.headers.authorization;
  if (
    typeof authorization !== "string" ||
    !authorization.startsWith("Bearer ") ||
    authorization.length === "Bearer ".length
  ) {
    response
      .writeHead(401, {
        "content-type": "text/plain; charset=utf-8",
        "www-authenticate": "Bearer",
      })
      .end("gateway internal identity required");
    return false;
  }
  try {
    verifyToken(authorization.slice("Bearer ".length));
    return true;
  } catch {
    response
      .writeHead(401, {
        "content-type": "text/plain; charset=utf-8",
        "www-authenticate": "Bearer error=\"invalid_token\"",
      })
      .end("gateway internal identity rejected");
    return false;
  }
}
